### ISSUE TITLE
NaN pseudo DICE with healty train and val loss

### ISSUE THREAD

**Issue by bastienchesaux** [MIC-DKFZ/LongiSeg#8]
Hello,

I have been playing around with nnUNet and LongiSeg on a dataset of micro-CT images of mice for binary segmentation. nnUNet training runs without any trouble, however, after having adapted the contents of the nnUNet_raw folder for compatibility with LongiSeg, I run into trouble with the values of pseudo Dice during training. Although the train and validation losses seem healthy, the pseudo dice is zero for three epochs and then goes to NaN for the rest of the training, after getting the following warnings:
`2026-03-27 13:17:07.450014: Epoch 3
2026-03-27 13:17:07.450090: Current learning rate: 0.00997
/scratch/chesaux/dev-longitudinal-tumour-segmentation-and-tracking/.venv/lib/python3.10/site-packages/numpy/_core/fromnumeric.py:86: RuntimeWarning: overflow encountered in reduce
  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
/scratch/chesaux/LongiSeg/longiseg/training/LongiSegTrainer/nnUNetTrainerNoLongi.py:1065: RuntimeWarning: invalid value encountered in scalar divide
  global_dc_per_class = [i for i in [2 * i / (2 * i + j + k) for i, j, k in zip(tp, fp, fn)]]
/scratch/chesaux/LongiSeg/longiseg/training/LongiSegTrainer/nnUNetTrainerNoLongi.py:1066: RuntimeWarning: Mean of empty slice
  mean_fg_dice = np.nanmean(global_dc_per_class)
2026-03-27 13:18:03.332421: train_loss -0.0287
2026-03-27 13:18:03.332512: val_loss -0.0734
2026-03-27 13:18:03.332553: Pseudo dice [np.float16(nan)]
2026-03-27 13:18:03.332623: Epoch time: 55.88 s
`
This behaviour is exactly the same if i use a different cross-validation fold.

progress.png:
<img width="3000" height="5400" alt="Image" src="https://github.com/user-attachments/assets/91f786bf-6639-4fff-b08e-e718cd380c10" />

The LongiSeg_plan_and_preprocess command with the --verify_Dataset_integrity flag ran without any error messages. I have also checked that the label files in the gt_segmentations folder do contain both classes (background and foreground) following the recommendations regarding a similar issue on nnUNet. 

The only modifications I have done to the dataset that works with nnUNet is to add a patientsTr.json file (see attached)  and remove cases when only one scan is available.

[patientsTr.json](https://github.com/user-attachments/files/26305311/patientsTr.json)

Note that all the images from the same case (patient) are the same shape, BUT I they are not co-registered yet. I want to benchmark the results without registration to then be able to compare different registration techniques, so I expect a bad result for this run.

I hope this is enough info to get an idea of where my problem could come from. Let me know if you need anything else.
Thank you for your help !

**Comment by Am0n22** [comment]
Hello,

I recently encountered and debugged this exact same issue. Your dataset, patientsTr.json should be correct.

The problem is line 563 of LongiSegTrainer.py (/longiseg/training/LongiSegTrainer/LongiSegTrainer.py):

`predicted_segmentation_onehot = torch.zeros(output.shape, device=output.device, dtype=torch.float16)
`

The label is transformed to torch.float16. During validation, the code calculates True Positives (TP), False Positives (FP), etc., by summing up the voxel counts. The maximum value a float16 tensor can hold is 65,504. Since your 3D images have way more voxels than that, the sum overflows to inf, which eventually propagates as NaN (as seen in your log: Pseudo dice [np.float16(nan)]).

To fix this, you just need to make modify torch.float16 into torch.float32. It works for me. Hope this also solves your problem.

**Comment by bastienchesaux** [comment]
Hi, thanks a lot for your answer, that was it! problem solved

**Comment by mrokuss** [comment]
Awesome, thanks for the catch! I just pushed a fix.

Best, Max
